### PR TITLE
Create a skeleton for the 'blobuploadprocessor'.

### PR DESCRIPTION
**Pull request by michaelsafyan**
#### Description

Create a skeleton for the `blobuploadprocessor` component.

A more fleshed out implementation exists in:

 - [`blobuploadconnector`](https://github.com/michaelsafyan/open-telemetry.opentelemetry-collector-contrib/tree/blob_upload_connector/connector/blobuploadconnector) - span + span events implementation
 - [`blobuploadconnector_logs`](https://github.com/michaelsafyan/open-telemetry.opentelemetry-collector-contrib/tree/blob_upload_connector_logs) - start of implementation for logs, branching from the span + span events one

However, the size of the entire thing is a bit unwieldy and likely to become a barrier for an effective code review.

In order to make progress upstreaming this and to ensure a reasonable implementation path, I'm now trying to break out little bits of the implementation into separate, smaller, more focused PRs. And, in the process of this, attempting to simplify the code, improve testing, etc. over the reference proof-of-concept.

#### Link to tracking issue

https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33737 : "New component: Blob Upload Processor"

#### Testing

Contains a "smoke test" in `package_test.go` that validates the basic contract for the connector and ensures non-crashing.

Also contains a few other unit tests and placeholders for additional unit tests.

As actual implementation pieces land in subsequent PRs, we will add additional corresponding tests.

#### Documentation

Contains a `README.md` that explains the purpose of this component and its intended trajectory.

Also contains a `metadata.yaml` providing structured documentation for the component.

We will update the `README.md` document as additional implementation lands to keep it up-to-date with actual state.

### CODE REVIEW

**Review by codeboten — COMMENTED** [review]
Pinging @dashpole as the sponsor for this component

**Comment by dashpole** [comment]
Ack.  My availability is limited until after kubecon, but I will try to get to this when I can.

File: connector/blobuploadconnector/metadata.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+type: blobuploadconnector
+scope_name: otelcol/blobuploadconnector
+
+status:
+  class: connector
```
Comment (dashpole) [inline connector/blobuploadconnector/metadata.yaml:5]: connectors are generally only used when we need to convert between different telemetry types (e.g. traces to logs).  Consider making this a processor unless we have use-cases where we need to convert between the signal types.

File: connector/blobuploadconnector/logs.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import (
+	"context"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/connector"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/plog"
+)
+
+type passThroughLogsConnector struct {
```
Comment (dashpole) [inline connector/blobuploadconnector/logs.go:15]: is this just meant to be a placeholder for the eventual implementation?  Consider using the name we will eventually use (e.g. logBlobUploader), to minimize future changes.

File: connector/blobuploadconnector/logs.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+	nextConsumer consumer.Logs
+	component.StartFunc
```
Comment (dashpole) [inline connector/blobuploadconnector/logs.go:17]: It is probably best to add start and shutdown functions to this struct, rather than add these functions to the struct, since calling them will produce a nil pointer exception.

File: connector/blobuploadconnector/metadata.yaml (hunk 1 above)
Comment (michaelsafyan) [reply connector/blobuploadconnector/metadata.yaml:5]: Thanks for the review @dashpole . 

My initial approach was to use a processor, but I ran into issues around data ownership (because I wanted to modify the data). Additionally, since this doesn't do pure computation but also does I/O to upload the data, it wasn't clear that a processor was the right fit. @jsuereth had suggested a connector as an approach to consider.

You can see some of the relevant history in https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33737 ; in particular, https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33737#issuecomment-2225861351 from July 12th shows where I was working on it as a processor (in a development branch named [`blob_writer_span_processor`](https://github.com/michaelsafyan/open-telemetry.opentelemetry-collector-contrib/tree/blob_writer_span_processor) ) before switching to the connector approach.

Is it OK with you if we proceed with the connector solution? [It sounds like this feedback was an optional rather than mandatory piece of feedback, but wanted to confirm].

File: connector/blobuploadconnector/logs.go (hunk 2 above)
Comment (michaelsafyan) [reply connector/blobuploadconnector/logs.go:15]: It is not a placeholder in that the full implementation will still include the `passthrough` variant.

You can see more clearly where this is going in the more fleshed out implementation sketch in:

 - [`[michaelsafyan/otel-contrib] connector/blobuploadconnector/traces.go`](https://github.com/michaelsafyan/open-telemetry.opentelemetry-collector-contrib/blob/blob_upload_connector/connector/blobuploadconnector/traces.go)

In particular:

```
func createTracesToTracesConnectorWithRegistryFactory(...) {
  ...
  if cfg.Traces == nil {
	  ...
	  return &passThroughTracesConnector{
		  nextConsumer: nextConsumer,
	  }, nil
  }
  ...
 result := &tracesToTracesImpl{ ... }
 ...
 return result, nil
}
```

Basically, the real implementation has some overhead even when there are no rules in effect, and so the factory function will instantiate a pass-through rather than a real implementation if the config does not have any rules which would be applied. This can be the case, for example, when configuration applies only to traces but not logs.

File: connector/blobuploadconnector/metadata.yaml (hunk 1 above)
Comment (dashpole) [reply connector/blobuploadconnector/metadata.yaml:5]: Processors can modify data (just make sure to set the "mutates data" capability to true).  If you do not have a cross-signal use case then this needs to be a processor.  This is blocking feedback.

File: connector/blobuploadconnector/logs.go (hunk 2 above)
Comment (dashpole) [reply connector/blobuploadconnector/logs.go:15]: You don't need to do this. If someone doesn't want this processor, they should omit it from their trace pipeline, rather than include it without configuration.  You should return an error if someone creates a blobuploader for traces without including trace configuration.

File: connector/blobuploadconnector/logs.go (hunk 3 above)
Comment (michaelsafyan) [reply connector/blobuploadconnector/logs.go]: Done.

File: connector/blobuploadconnector/logs.go (hunk 2 above)
Comment (michaelsafyan) [reply connector/blobuploadconnector/logs.go]: Seems reasonable.

I've renamed these so that they are placeholders for the eventual, real implementation.

When I build on this to upstream the other pieces of the implementation, I will be stricter in the validation rather than attempting to fail more gracefully.

File: connector/blobuploadconnector/metadata.yaml (hunk 1 above)
Comment (michaelsafyan) [reply connector/blobuploadconnector/metadata.yaml]: Thanks. I've updated the code to use a processor.

File: .chloggen/blobuploadprocessor.yaml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+change_type: new_component
+component: blobuploadprocessor
+note: Processor for uploading pieces of a signal to blob storage and replacing with a ref
+issues: [33737]
+subtext: Skeleton implementation at the moment.
+change_logs: []
```
Comment (dashpole) [inline .chloggen/blobuploadprocessor.yaml:6]:
```suggestion
change_logs: [api, user]
```

File: .github/ISSUE_TEMPLATE/feature_request.yaml [hunk 5]
```diff
@@ -20,6 +20,7 @@ body:
       - confmap/provider/aesprovider
       - confmap/provider/s3provider
       - confmap/provider/secretsmanagerprovider
+      - connector/blobupload
```
Comment (dashpole) [inline .github/ISSUE_TEMPLATE/feature_request.yaml:23]: update to processor

File: .github/ISSUE_TEMPLATE/other.yaml [hunk 6]
```diff
@@ -20,6 +20,7 @@ body:
       - confmap/provider/aesprovider
       - confmap/provider/s3provider
       - confmap/provider/secretsmanagerprovider
+      - connector/blobupload
```
Comment (dashpole) [inline .github/ISSUE_TEMPLATE/other.yaml:23]: update to processor

File: .github/ISSUE_TEMPLATE/unmaintained.yaml [hunk 7]
```diff
@@ -25,6 +25,7 @@ body:
       - confmap/provider/aesprovider
       - confmap/provider/s3provider
       - confmap/provider/secretsmanagerprovider
+      - connector/blobupload
```
Comment (dashpole) [inline .github/ISSUE_TEMPLATE/unmaintained.yaml:28]: update to processor

File: processor/blobuploadprocessor/README.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
+# Blob Upload Connector
```
Comment (dashpole) [inline processor/blobuploadprocessor/README.md:1]:
```suggestion
# Blob Upload Processor
```

File: processor/blobuploadprocessor/factory.go [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+package blobuploadprocessor // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/blobuploadprocessor"
+
+import (
+	"go.opentelemetry.io/collector/processor"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/blobuploadprocessor/internal/metadata"
+)
+
+// NewFactoryWithDeps instantiates the factory with dependency injection, allowing
```
Comment (dashpole) [inline processor/blobuploadprocessor/factory.go:12]:
```suggestion
// newFactoryWithDeps instantiates the factory with dependency injection, allowing
```

File: processor/blobuploadprocessor/metadata.yaml [hunk 10]
```diff
@@ -0,0 +1,21 @@
+type: blobuploadconnector
+scope_name: otelcol/blobuploadconnector
+
+status:
+  class: processor
+  stability:
+    # Implementation still in-progress. Please follow along at:
+    # https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33737
+    development: [traces, logs]
+  distributions: [contrib]
```
Comment (dashpole) [inline processor/blobuploadprocessor/metadata.yaml:10]: adding this to contrib is done after the implementation is complete (it reaches alpha).
```suggestion
  distributions: []
```

File: processor/blobuploadprocessor/metadata.yaml [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+  codeowners:
+    active: [michaelsafyan, dashpole, braydonk]
+
+tests:
+  config:
+  goleak:
+    skip: false
+    ignore:
+      # See https://github.com/census-instrumentation/opencensus-go/issues/1191 for more information.
+      top: go.opencensus.io/stats/view.(*worker).start
```
Comment (dashpole) [inline processor/blobuploadprocessor/metadata.yaml:20]: Strange that this is required given opencensus isn't in the go.sum. We can try removing it in a follow-up.

**Review by dashpole — COMMENTED** [review]
looking good.  Just small tweaks
